src/RtpFecQueue.c

```diff
@@ -0,0 +1,291 @@
+                    dataOffset += 4; // 2 additional fields
+                }
+
+                PNV_VIDEO_PACKET nvPacket = (PNV_VIDEO_PACKET)(((char*)rtpPacket) + dataOffset);
+                nvPacket->frameIndex = queue->currentFrameNumber;
+                
+                //Remove padding of generated packet
+                int size = StreamConfig.packetSize + dataOffset;
+                if (rtpPacket->sequenceNumber == ushort(queue->bufferLowestSequenceNumber + queue->bufferDataPackets - 1)) {
+                    while (packets[i][size-1] == 0) {
+                        size--;
+                    }
+                }
+
+                memcpy(&packets[i][receiveSize], &size, sizeof(int));
+                queuePacket(queue, queueEntry, 0, rtpPacket);
+            } else {
+                free(packets[i]);
+            }
+        }
+    }
+
+    cleanup:
+    if (missing != NULL)
+        free(missing);
+
+    if (packets != NULL)
+        free(packets);
+
+    if (marks != NULL)
+        free(marks);
+}
+
+static void removeEntry(PRTP_FEC_QUEUE queue, PRTPFEC_QUEUE_ENTRY entry) {
+    LC_ASSERT(entry != NULL);
+    LC_ASSERT(queue->queueSize > 0);
+    LC_ASSERT(queue->readyQueueHead != NULL);
+    LC_ASSERT(queue->readyQueueTail != NULL);
+
+    if (queue->queueHead == entry) {
+        queue->queueHead = entry->next;
+    }
+    if (queue->queueTail == entry) {
+        queue->queueTail = entry->prev;
+    }
+
+    if (entry->prev != NULL) {
+        entry->prev->next = entry->next;
+    }
+    if (entry->next != NULL) {
+        entry->next->prev = entry->prev;
+    }
+    queue->queueSize--;
+}
+
+int RtpfAddPacket(PRTP_FEC_QUEUE queue, PRTP_PACKET packet, PRTPFEC_QUEUE_ENTRY packetEntry) {
+    if (isBefore(packet->sequenceNumber, queue->nextRtpSequenceNumber)) {
+        // Reject packets behind our current sequence number
+        return RTPF_RET_REJECTED;
+    }
+
+    int dataOffset = sizeof(*packet);
+    if (packet->header & FLAG_EXTENSION) {
+        dataOffset += 4; // 2 additional fields
+    }
+
+    PNV_VIDEO_PACKET nvPacket = (PNV_VIDEO_PACKET)(((char*)packet) + dataOffset);
+
+    if (nvPacket->frameIndex != queue->currentFrameNumber) {
+        //Make current frame available to depacketizer
+        queue->currentFrameNumber = nvPacket->frameIndex;
+        if (queue->queueTail == NULL) {
+            queue->queueHead = queue->bufferHead;
+            queue->queueTail = queue->bufferTail;
+        } else {
+            queue->bufferTail->next = queue->bufferHead;
+            queue->queueTail = queue->bufferTail;
+        }
+        queue->bufferHead = NULL;
+        queue->bufferTail = NULL;
+
+        queue->queueSize = queue->bufferSize;
+        queue->nextRtpSequenceNumber = queue->bufferHighestSequenceNumber;
+        
+        int fecIndex = (nvPacket->fecInfo & 0xFF000) >> 12;
+        queue->bufferLowestSequenceNumber = packet->sequenceNumber - fecIndex;
```

Comment:
What happens if the first packet in a frame is dropped? Would this make `queue->bufferLowestSequenceNumber` wrong?